Source/WebCore/rendering/svg/RenderSVGResourcePattern.cpp

```diff
@@ -0,0 +1,231 @@
+
+    const auto& svgStyle = style.svgStyle();
+
+    context.setAlpha(svgStyle.strokeOpacity());
+    SVGRenderSupport::applyStrokeStyleToContext(context, style, targetRenderer);
+    if (svgStyle.vectorEffect() == VectorEffect::NonScalingStroke) {
+        if (is<RenderSVGShape>(targetRenderer))
+            pattern->setPatternSpaceTransform(transformOnNonScalingStroke(targetRenderer, pattern->patternSpaceTransform()));
+    }
+    context.setStrokePattern(pattern.releaseNonNull());
+    return true;
+}
+
+bool RenderSVGResourcePattern::buildTileImageTransform(const RenderElement& renderer, const PatternAttributes& attributes, const SVGPatternElement& patternElement, FloatRect& patternBoundaries, AffineTransform& tileImageTransform) const
+{
+    auto objectBoundingBox = renderer.objectBoundingBox();
+    patternBoundaries = calculatePatternBoundaries(attributes, objectBoundingBox, patternElement);
+    if (patternBoundaries.width() <= 0 || patternBoundaries.height() <= 0)
+        return false;
+
+    auto viewBoxCTM = SVGFitToViewBox::viewBoxToViewTransform(attributes.viewBox(), attributes.preserveAspectRatio(), patternBoundaries.width(), patternBoundaries.height());
+
+    // Apply viewBox/objectBoundingBox transformations.
+    if (!viewBoxCTM.isIdentity())
+        tileImageTransform = viewBoxCTM;
+    else if (attributes.patternContentUnits() == SVGUnitTypes::SVG_UNIT_TYPE_OBJECTBOUNDINGBOX)
+        tileImageTransform.scale(objectBoundingBox.width(), objectBoundingBox.height());
+
+    return true;
+}
+
+RefPtr<ImageBuffer> RenderSVGResourcePattern::createTileImage(const PatternAttributes& attributes, const FloatRect& tileBoundaries, const FloatRect& absoluteTileBoundaries, const AffineTransform& tileImageTransform) const
+{
+    auto clampedAbsoluteTileBoundaries = ImageBuffer::clampedRect(absoluteTileBoundaries);
+    auto tileImage = ImageBuffer::create(roundedIntSize(clampedAbsoluteTileBoundaries.size()), RenderingPurpose::Unspecified, 1, DestinationColorSpace::SRGB(), PixelFormat::BGRA8);
+    if (!tileImage)
+        return nullptr;
+
+    auto& tileImageContext = tileImage->context();
+
+    GraphicsContextStateSaver stateSaver(tileImageContext);
+
+    FloatSize unclampedSize = roundedIntSize(tileBoundaries.size());
+
+    // Compensate rounding effects, as the absolute target rect is using floating-point numbers and the image buffer size is integer.
+    tileImageContext.scale(unclampedSize / tileBoundaries.size());
+
+    // The image buffer represents the final rendered size, so the content has to be scaled (to avoid pixelation).
+    tileImageContext.scale(clampedAbsoluteTileBoundaries.size() / tileBoundaries.size());
+
+    auto* patternRenderer = static_cast<RenderSVGResourcePattern*>(attributes.patternContentElement()->renderer());
+    ASSERT(patternRenderer);
+    ASSERT(patternRenderer->hasLayer());
+
+    // We only want to detect cycles for patternRenderer != this, since paintSVGResourceLayer itself
```

Comment:
When does that happen? referenced patterns via href?

Reply:
It seemed to happen in Debug with a simple case, like svg/repaint/pattern-object-bounding-box-transformed.html.